File: DailyStudent/src/main/java/pt/uc/cm/daily_student/activities/GlobalNotes.java
package pt.uc.cm.daily_student.activities;

import android.app.AlertDialog;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.NotificationCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.ContextMenu;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.SimpleCursorAdapter;
import android.widget.Toast;

import pt.uc.cm.daily_student.R;
import pt.uc.cm.daily_student.adapters.GlobalNotesDbAdapter;
import pt.uc.cm.daily_student.fragments.NoteEdit;
import pt.uc.cm.daily_student.interfaces.INetworkReceiveEvents;
import pt.uc.cm.daily_student.models.MessagePacket;
import pt.uc.cm.daily_student.utils.NetworkUtils;

public class GlobalNotes extends StructureActivity {
    private final String TAG = GlobalNotes.class.getSimpleName();

    private static final int ACTIVITY_CREATE = 0;
    private static final int ACTIVITY_EDIT = 1;

    private static final int DELETE_ID = Menu.FIRST + 1;
    private static final int SHARE_ID = Menu.FIRST + 2;

    private GlobalNotesDbAdapter mGlobalNotesDbAdapter;
    private Cursor mNotesCursor;

    private ListView lvGlobalNotes;
    private long id;
    private String author = null, ip = null;

    private ProgressDialog progressDialog = null;

    int id_noti = 0;

    @Override
    protected void onResume() {
        super.onResume();


    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        readPreferencesUser();
        setContentView(R.layout.activity_global_notes);
        setTitle(getString(R.string.DayliStudentActivitySharedNote));

        //abrir a base de dados das notas textuais
        mGlobalNotesDbAdapter = new GlobalNotesDbAdapter(this);
        mGlobalNotesDbAdapter.open();

        fillData();

        id = 0;

        lvGlobalNotes = findViewById(R.id.lv_global_notes);
        registerForContextMenu(lvGlobalNotes);

        lvGlobalNotes.setOnItemClickListener((adapterView, view, position, id) -> {
            Cursor c = mNotesCursor;
            c.moveToPosition(position);
            Intent i = new Intent(getApplicationContext(), NoteEdit.class);
            i.putExtra(GlobalNotesDbAdapter.KEY_ROWID, id);
            i.putExtra(GlobalNotesDbAdapter.KEY_AUTHOR, c.getString(c.getColumnIndexOrThrow(GlobalNotesDbAdapter.KEY_AUTHOR)));
            i.putExtra(GlobalNotesDbAdapter.KEY_TITLE, c.getString(c.getColumnIndexOrThrow(GlobalNotesDbAdapter.KEY_TITLE)));
            i.putExtra(GlobalNotesDbAdapter.KEY_BODY, c.getString(c.getColumnIndexOrThrow(GlobalNotesDbAdapter.KEY_BODY)));
            startActivityForResult(i, ACTIVITY_EDIT);
        });
    }

    private void fillData() {
        // Get all of the rows from the database and create the item list
        mNotesCursor = mGlobalNotesDbAdapter.fetchAllGlobalNotes();
        startManagingCursor(mNotesCursor);

        // Create an array to specify the fields we want to display in the list (only TITLE)
        String[] from = new String[]{GlobalNotesDbAdapter.KEY_AUTHOR, GlobalNotesDbAdapter.KEY_TITLE, GlobalNotesDbAdapter.KEY_BODY, GlobalNotesDbAdapter.KEY_DATE};

        // and an array of the fields we want to bind those fields to (in this case just text1)
        int[] to = new int[]{R.id.tvAuthor, R.id.tvGlobalTitulo, R.id.tvGlobalBody, R.id.tvGlobalDate};

        // Now create a simple cursor adapter and set it to display
        SimpleCursorAdapter notes =
                new SimpleCursorAdapter(this, R.layout.notes_global_row, mNotesCursor, from, to);
        lvGlobalNotes = findViewById(R.id.lv_global_notes);
        lvGlobalNotes.setAdapter(notes);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_sup_globa, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public void onCreateContextMenu(ContextMenu menu, View v, ContextMenu.ContextMenuInfo menuInfo) {
        super.onCreateContextMenu(menu, v, menuInfo);

        menu.add(0, DELETE_ID, 0, R.string.DayliStudentActivityDeleteNote);
        menu.add(0, SHARE_ID, 0, R.string.DayliStudentActivityShareNote);
    }

    @Override
    public boolean onContextItemSelected(MenuItem item) {
        AdapterView.AdapterContextMenuInfo info = (AdapterView.AdapterContextMenuInfo) item.getMenuInfo();
        switch (item.getItemId()) {
            case DELETE_ID:
                mGlobalNotesDbAdapter.deleteGlobalNote(info.id);
                fillData();
                return true;
            case SHARE_ID:
                senderDialog();
                id = info.id;
                return true;
        }
        return super.onContextItemSelected(item);
    }

    public void createBudgetAction(MenuItem item) {
        startActivityForResult(new Intent(this, NoteEdit.class), ACTIVITY_CREATE);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent intent) {
        super.onActivityResult(requestCode, resultCode, intent);

        if (resultCode == RESULT_OK) {
            Bundle extras = intent.getExtras();

            switch (requestCode) {
                case ACTIVITY_CREATE:
                    String title = extras.getString(GlobalNotesDbAdapter.KEY_TITLE);
                    String body = extras.getString(GlobalNotesDbAdapter.KEY_BODY);
                    mGlobalNotesDbAdapter.createGlobalNote(author, title, body);
                    fillData();
                    break;
                case ACTIVITY_EDIT:
                    Integer mRowId = extras.getInt(GlobalNotesDbAdapter.KEY_ROWID);
                    if (mRowId != null) {
                        String editTitle = extras.getString(GlobalNotesDbAdapter.KEY_TITLE);
                        String editBody = extras.getString(GlobalNotesDbAdapter.KEY_BODY);
                        mGlobalNotesDbAdapter.updateGlobalNote(mRowId, author, editTitle, editBody);
                    }
                    fillData();
                    break;
                default:
            }
        }
    }

    private void senderDialog() {
        final EditText edtIP = new EditText(this);

        // TODO: implement get IP from QRCode
        edtIP.setText(ip);
        AlertDialog ad = new AlertDialog.Builder(this)
                .setTitle(R.string.DayliStudentActivitySendingSharedNote)
                .setMessage(R.string.DayliStudentActivityIPSendingSharedNote)
                .setView(edtIP)
                .setPositiveButton(R.string.DayliStudentActivityButtonSendingSharedNote, (dialogInterface, i) -> {
                    // TODO: check for permissions Internet
                    NetworkUtils networkUtils = NetworkUtils.getInstance();
                    networkUtils.sendData(mGlobalNotesDbAdapter.getGlobalNoteToSend(this.id),
                            this::posDataSend);
                })
                .setOnCancelListener(dialogInterface -> {
                })
                .create();
        ad.show();
    }

    public void onReceiveNoteClick(MenuItem item) {
        NetworkUtils networkUtils = NetworkUtils.getInstance();
        networkUtils.receiveData(new INetworkReceiveEvents() {
            @Override
            public void onPreDataReceived() {
                preDataReceived();
            }

            @Override
            public void onPosDataReceived(MessagePacket messagePacket) {
                posDataReceived(messagePacket);
            }
        });
    }

    public void receiveNote(MenuItem item) {
        String ip = getLocalIpAddress();
        pd = new ProgressDialog(this);
        pd.setMessage(getString(R.string.waitingConection) + "\n(IP: " + ip
                + ")");
        pd.setTitle(getString(R.string.receive));
        pd.setOnCancelListener(dialog -> {
            if (receiverSocket != null) {
                try {
                    receiverSocket.close();
                } catch (IOException ignored) {
                }
                receiverSocket = null;
            }
        });
        pd.show();

        Thread t = new Thread(new waitConnection());
        t.start();
    }

    //TODO: CONVERT TO ASYNC
    public class waitConnection implements Runnable {
        @Override
        public void run() {
            try {
                receiverSocket = new ServerSocket(9001);
                senderSocket = receiverSocket.accept();
                receiverSocket.close();
                receiverSocket = null;
                Thread t = new Thread(new receiveMSG());
                t.start();
            } catch (Exception e) {
                e.printStackTrace();
                senderSocket = null;
            }
            procMsg.post(() -> {
                pd.dismiss();
                if (receiverSocket != null)
                    try {
                        receiverSocket.close();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
            });
        }
    }

    //TODO: CONVERT TO ASYNC(2)
    public class receiveMSG implements Runnable {

        public void run() {
            try {
                inObj = new ObjectInputStream(senderSocket.getInputStream());

                final MessagePacket read = (MessagePacket) (inObj.readObject());

                procMsg.post(() -> {
                    mDbHelper.createGlobalNote(read.getAuthor(), read.getTitle(), read.getObs());

                    fillData();

                    Toast.makeText(getApplicationContext(), getString(R.string.DayliStudentActivityReceivedSharedNote) +
                            read.getTitle() + getString(R.string.DayliStudentActivityReceived2SharedNote) + read.getAuthor() + ".", Toast.LENGTH_LONG).show();

                    buildNotification(read);
                });
            } catch (Exception ex) {
                ex.printStackTrace();
                procMsg.post(() -> {
                    Toast.makeText(getApplicationContext(), "ERROR", Toast.LENGTH_LONG).show();
                    // Refresh main activity upon close of dialog box
                    Intent refresh = new Intent(getApplicationContext(), GlobalNotes.class);
                    startActivity(refresh);
                    finish();
                });
            }
        }
    }

    public class sendMSG implements Runnable {
        MessagePacket msg;

        public void run() {
            try {
                outObj = new ObjectOutputStream(senderSocket.getOutputStream());

                msg = mDbHelper.getGlobalNoteToSend(id);

                id = 0;

                outObj.writeObject(msg);
                outObj.flush();

                procMsg.post(() -> Toast.makeText(getApplicationContext(),
                        getString(R.string.DayliStudentActivitySuccessfullSendingSharedNote) + msg.getTitle() + " ]", Toast.LENGTH_LONG).show());
            } catch (Exception ex) {
                ex.printStackTrace();
                procMsg.post(() -> {
                });
            }
        }
    }

    public static String getLocalIpAddress() {
        try {
            for (Enumeration<NetworkInterface> en = NetworkInterface
                    .getNetworkInterfaces(); en.hasMoreElements(); ) {
                NetworkInterface intf = en.nextElement();
                for (Enumeration<InetAddress> enumIpAddr = intf
                        .getInetAddresses(); enumIpAddr.hasMoreElements(); ) {
                    InetAddress inetAddress = enumIpAddr.nextElement();
                    if (!inetAddress.isLoopbackAddress()
                            && inetAddress instanceof Inet4Address) {
                        return inetAddress.getHostAddress();
                    }
                }
            }
        } catch (SocketException ex) {
            ex.printStackTrace();
        }
        return null;
    }

    public void buildNotification(MessagePacket msg) {
        NotificationCompat.Builder mBuilder =
                new NotificationCompat.Builder(getApplicationContext())
                        .setSmallIcon(R.mipmap.ic_launcher)
                        .setContentTitle(msg.getAuthor())
                        .setContentText(msg.getTitle());
        Intent resultIntent = new Intent(getApplicationContext(), GlobalNotes.class);
        PendingIntent resultPendingIntent =
                PendingIntent.getActivity(
                        getApplicationContext(),
                        0,
                        resultIntent,
                        PendingIntent.FLAG_UPDATE_CURRENT
                );
        mBuilder.setContentIntent(resultPendingIntent);

        id_noti++;
        NotificationManager mNotifyMgr =
                (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
        mNotifyMgr.notify(id_noti, mBuilder.build());
    }

    private void preDataReceived() {
        // TODO: implement show IP from QRCode
        progressDialog = new ProgressDialog(this);
        progressDialog.setMessage(getString(R.string.waitingConection) + "\n(IP: " + ip + ")");
        progressDialog.setTitle(getString(R.string.receive));
        /* TODO: remove this if cancel works!
        progressDialog.setOnCancelListener(dialog -> {
            if (receiverSocket != null) {
                try {
                    receiverSocket.close();
                } catch (IOException ignored) {
                }
                receiverSocket = null;
            }
        });*/
        progressDialog.show();
    }

    private void posDataReceived(MessagePacket messagePacket) {
        if (messagePacket != null) {
            progressDialog.dismiss();

            mGlobalNotesDbAdapter.createGlobalNote(messagePacket.getAuthor(),
                    messagePacket.getTitle(),
                    messagePacket.getObs());

            fillData();

            Toast.makeText(getApplicationContext(),
                    getString(R.string.DayliStudentActivityReceivedSharedNote) +
                            messagePacket.getTitle()
                            + getString(R.string.DayliStudentActivityReceived2SharedNote)
                            + messagePacket.getAuthor() + ".", Toast.LENGTH_LONG).show();

            buildNotification(messagePacket);
        } else {
            Toast.makeText(getApplicationContext(), "ERROR", Toast.LENGTH_LONG).show();
            // Refresh main activity upon close of dialog box
            Intent refresh = new Intent(getApplicationContext(), GlobalNotes.class);
            startActivity(refresh);
            finish();
        }
    }

    private void posDataSend(MessagePacket messagePacket, boolean success) {
        Context context = getApplicationContext();
        Toast toast = success ?
                Toast.makeText(context, context.getString(R.string.SendingSharedNoteSuccess)
                        + ": " + messagePacket.getTitle(), Toast.LENGTH_LONG)
                :
                Toast.makeText(context,
                        context.getString(R.string.SendingSharedNoteFail)
                                + ": " + messagePacket.getTitle()
                                + context.getString(R.string.PleaseTryAgainLater),
                        Toast.LENGTH_LONG);
        toast.show();
    }
}

File: DailyStudent/src/main/java/pt/uc/cm/daily_student/activities/GlobalNotes.kt
package pt.uc.cm.daily_student.activities

import android.app.*
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.database.Cursor
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v4.app.NotificationCompat
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.ContextMenu
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ImageView
import android.widget.SimpleCursorAdapter
import android.widget.Toast
import com.google.zxing.qrcode.decoder.ErrorCorrectionLevel
import kotlinx.android.synthetic.main.activity_global_notes.*
import pt.uc.cm.daily_student.R
import pt.uc.cm.daily_student.adapters.GlobalNotesDbAdapter
import pt.uc.cm.daily_student.fragments.NoteEdit
import pt.uc.cm.daily_student.fragments.ScanQrCodeFragment
import pt.uc.cm.daily_student.interfaces.INetworkReceiveEvents
import pt.uc.cm.daily_student.interfaces.INetworkSendEvents
import pt.uc.cm.daily_student.interfaces.IScanQrCodeEvents
import pt.uc.cm.daily_student.models.MessagePacket
import pt.uc.cm.daily_student.utils.EncryptionUtils
import pt.uc.cm.daily_student.utils.NetworkUtils
import pt.uc.cm.daily_student.utils.QRCodeUtils

class GlobalNotes : AppCompatActivity() {

    private lateinit var mGlobalNotesDbAdapter: GlobalNotesDbAdapter
    private lateinit var mNotesCursor: Cursor

    private var id: Long = 0
    private var author: String? = null
    private var ip: String? = null

    private val progressDialog: ProgressDialog? = null

    private lateinit var sharedPreferences: SharedPreferences
    private var id_noti = 0

    companion object {
        private val TAG = GlobalNotes::class.java.simpleName

        private const val ACTIVITY_CREATE = 0
        private const val ACTIVITY_EDIT = 1

        private const val DELETE_ID = Menu.FIRST + 1
        private const val SHARE_ID = Menu.FIRST + 2
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        readPreferencesUser()
        setContentView(R.layout.activity_global_notes)
        title = getString(R.string.DayliStudentActivitySharedNote)

        // Open the text notes database
        mGlobalNotesDbAdapter = GlobalNotesDbAdapter(this)
        mGlobalNotesDbAdapter.open()

        fillData()

        id = 0

        registerForContextMenu(lvGlobalNotes)
        lvGlobalNotes!!.setOnItemClickListener { _, _, position, id ->
            mNotesCursor.moveToPosition(position)
            val i = Intent(applicationContext, NoteEdit::class.java)
            i.putExtra(GlobalNotesDbAdapter.KEY_ROWID, id)
            i.putExtra(GlobalNotesDbAdapter.KEY_AUTHOR, mNotesCursor.getString(mNotesCursor.getColumnIndexOrThrow(GlobalNotesDbAdapter.KEY_AUTHOR)))
            i.putExtra(GlobalNotesDbAdapter.KEY_TITLE, mNotesCursor.getString(mNotesCursor.getColumnIndexOrThrow(GlobalNotesDbAdapter.KEY_TITLE)))
            i.putExtra(GlobalNotesDbAdapter.KEY_BODY, mNotesCursor.getString(mNotesCursor.getColumnIndexOrThrow(GlobalNotesDbAdapter.KEY_BODY)))
            startActivityForResult(i, ACTIVITY_EDIT)
        }
    }

    override fun onResume() {
        super.onResume()

        // TODO: check Internet and Camera permissions?
    }

    private fun fillData() {
        // Get all of the rows from the database and create the item list
        mNotesCursor = mGlobalNotesDbAdapter.fetchAllGlobalNotes()
        startManagingCursor(mNotesCursor)

        // Create an array to specify the fields we want to display in the list (only TITLE)
        val from = arrayOf(GlobalNotesDbAdapter.KEY_AUTHOR, GlobalNotesDbAdapter.KEY_TITLE, GlobalNotesDbAdapter.KEY_BODY, GlobalNotesDbAdapter.KEY_DATE)

        // and an array of the fields we want to bind those fields to (in this case just text1)
        val to = intArrayOf(R.id.tvAuthor, R.id.tvGlobalTitulo, R.id.tvGlobalBody, R.id.tvGlobalDate)

        // Now create a simple cursor adapter and set it to display
        val notes = SimpleCursorAdapter(this, R.layout.notes_global_row, mNotesCursor, from, to)
        lvGlobalNotes!!.adapter = notes
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_sup_globa, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenu.ContextMenuInfo) {
        super.onCreateContextMenu(menu, v, menuInfo)

        menu.add(0, DELETE_ID, 0, R.string.DayliStudentActivityDeleteNote)
        menu.add(0, SHARE_ID, 0, R.string.DayliStudentActivityShareNote)
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        val info = item.menuInfo as AdapterView.AdapterContextMenuInfo
        when (item.itemId) {
            DELETE_ID -> {
                mGlobalNotesDbAdapter.deleteGlobalNote(info.id)
                fillData()
                return true
            }
            SHARE_ID -> {
                senderDialog()
                id = info.id
                return true
            }
        }
        return super.onContextItemSelected(item)
    }

    fun createBudgetAction(item: MenuItem) {
        startActivityForResult(Intent(this, NoteEdit::class.java), ACTIVITY_CREATE)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
        super.onActivityResult(requestCode, resultCode, intent)

        if (resultCode == Activity.RESULT_OK) {
            val extras = intent!!.extras

            when (requestCode) {
                ACTIVITY_CREATE -> {
                    val title = extras!!.getString(GlobalNotesDbAdapter.KEY_TITLE)
                    val body = extras.getString(GlobalNotesDbAdapter.KEY_BODY)
                    mGlobalNotesDbAdapter.createGlobalNote(author, title, body)
                    fillData()
                }
                ACTIVITY_EDIT -> {
                    val mRowId = extras!!.getInt(GlobalNotesDbAdapter.KEY_ROWID)
                    val editTitle = extras.getString(GlobalNotesDbAdapter.KEY_TITLE)
                    val editBody = extras.getString(GlobalNotesDbAdapter.KEY_BODY)
                    mGlobalNotesDbAdapter.updateGlobalNote(mRowId.toLong(), author, editTitle, editBody)
                    fillData()
                }
            }
        }
    }

    private fun senderDialog() {
        val scanQrCodeFragment = ScanQrCodeFragment()
        scanQrCodeFragment.title = getString(R.string.DayliStudentActivitySendingSharedNote)
        scanQrCodeFragment.scanQrCodeEvents = object : IScanQrCodeEvents {
            override fun onDataScanned(scannedText: String) {
                val decryptedIp = EncryptionUtils.instance.getDecryptionString(scannedText)

                Toast.makeText(applicationContext, decryptedIp, Toast.LENGTH_LONG).show()

                val networkUtils = NetworkUtils.instance
                networkUtils.sendData(decryptedIp,
                        mGlobalNotesDbAdapter.getGlobalNoteToSend(id), object : INetworkSendEvents {
                    override fun onPosDataSend(messagePacket: MessagePacket, success: Boolean?) {
                        posDataSend(messagePacket, success!!)
                    }
                })
            }
        }
        scanQrCodeFragment.show(supportFragmentManager, null)
    }

    fun onReceiveNoteClick(item: MenuItem) {
        val networkUtils = NetworkUtils.instance
        networkUtils.receiveData(object : INetworkReceiveEvents {
            override fun onPreDataReceived() {
                preDataReceived()
            }

            override fun onPosDataReceived(messagePacket: MessagePacket) {
                posDataReceived(messagePacket)
            }
        })
    }

    private fun readPreferencesUser() {
        // TODO: check why textSize is never used
        var textSize = -1
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this@GlobalNotes)

        author = sharedPreferences.getString("nameKey", "DEFAULT")
        ip = sharedPreferences.getString("serverKey", "0.0.0.0")

        when (sharedPreferences.getString("themeKey", "YellowTheme")) {
            "RedTheme" -> setTheme(R.style.RedTheme)
            "YellowTheme" -> setTheme(R.style.YellowTheme)
            "GreenTheme" -> setTheme(R.style.GreenTheme)
        }

        Log.i(TAG, "selected size: " + sharedPreferences.getString("fontSizeKey", "darkab")!!)
        when (sharedPreferences.getString("fontSizeKey", "normal")) {
            "smallest" -> textSize = 12
            "small" -> textSize = 14
            "normal" -> textSize = 16
            "large" -> textSize = 18
            "largest" -> textSize = 20
        }
    }

    private fun buildNotification(msg: MessagePacket) {
        val mBuilder = NotificationCompat.Builder(applicationContext)
                .setSmallIcon(R.mipmap.ic_launcher)
                .setContentTitle(msg.author)
                .setContentText(msg.title)
        val resultIntent = Intent(applicationContext, GlobalNotes::class.java)
        val resultPendingIntent = PendingIntent.getActivity(
                applicationContext,
                0,
                resultIntent,
                PendingIntent.FLAG_UPDATE_CURRENT
        )
        mBuilder.setContentIntent(resultPendingIntent)

        id_noti++
        val mNotifyMgr = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        mNotifyMgr.notify(id_noti, mBuilder.build())
    }

    private fun preDataReceived() {
        val localIp = NetworkUtils.localIpAddress
        val encryptedLocalIp = EncryptionUtils.instance.encryptionString(localIp!!)?.encryptMsg()
        val qrCodeBitmap = QRCodeUtils.newInstance(this)
                .setContent(encryptedLocalIp)
                .setErrorCorrectionLevel(ErrorCorrectionLevel.Q).setMargin(2)
                .qrcOde

        val alertDialog = layoutInflater.inflate(R.layout.receiver_alert_dialog, null)
        val imageView = alertDialog.findViewById<ImageView>(R.id.qrCode)
        imageView.setImageBitmap(qrCodeBitmap)

        val dialog = AlertDialog.Builder(this)
        dialog.setTitle(getString(R.string.receive))
        dialog.setView(alertDialog)
        dialog.setNegativeButton(getString(R.string.cancel)) { _, _ -> NetworkUtils.instance.stopAll() }
        dialog.show()
    }

    private fun posDataReceived(messagePacket: MessagePacket?) {
        if (messagePacket != null) {
            progressDialog!!.dismiss()

            mGlobalNotesDbAdapter.createGlobalNote(messagePacket.author,
                    messagePacket.title,
                    messagePacket.obs)

            fillData()

            Toast.makeText(applicationContext,
                    getString(R.string.DayliStudentActivityReceivedSharedNote) +
                            messagePacket.title
                            + getString(R.string.DayliStudentActivityReceived2SharedNote)
                            + messagePacket.author + ".", Toast.LENGTH_LONG).show()

            buildNotification(messagePacket)
        } else {
            Toast.makeText(applicationContext, getString(R.string.error), Toast.LENGTH_LONG).show()
            // Refresh main activity upon close of dialog box
            val refresh = Intent(applicationContext, GlobalNotes::class.java)
            startActivity(refresh)
            finish()
        }
    }

    private fun posDataSend(messagePacket: MessagePacket, success: Boolean) {
        val context = applicationContext
        val toast = if (success)
            Toast.makeText(context, context.getString(R.string.SendingSharedNoteSuccess)
                    + ": " + messagePacket.title, Toast.LENGTH_LONG)
        else
            Toast.makeText(context,
                    context.getString(R.string.SendingSharedNoteFail)
                            + ": " + messagePacket.title
                            + context.getString(R.string.PleaseTryAgainLater),
                    Toast.LENGTH_LONG)
        toast.show()
    }
}
